### PR TITLE
make array.reshape compatible with numpy

### PR DESCRIPTION
**Pull request by szha**
## Description ##
Allow n ints as input to array.reshape.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Allow n ints as input to array.reshape.

### CODE REVIEW

File: python/mxnet/ndarray/ndarray.py [hunk 1]
```diff
@@ -926,12 +926,12 @@ def _at(self, idx):
             self.handle, mx_uint(idx), ctypes.byref(handle)))
         return NDArray(handle=handle, writable=self.writable)
 
-    def reshape(self, shape):
+    def reshape(self, *shape, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: (self, *args, shape=None)?

File: python/mxnet/ndarray/ndarray.py [hunk 2]
```diff
@@ -968,6 +973,14 @@ def reshape(self, shape):
         array([[-1., -1., -1.],
                [-1., -1., -1.]], dtype=float32)
         """
+        if len(shape) == 1 and isinstance(shape[0], (list, tuple)):
+            shape = shape[0]
+        elif not len(shape):
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: what happens for reshape(1, 2, shape=1)?

File: python/mxnet/ndarray/ndarray.py (hunk 1 above)
Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: This doesn't work in py2

File: python/mxnet/ndarray/ndarray.py (hunk 2 above)
Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: keyword argument is ignored.

Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: would that be a problem?

Comment (piiswrong) [reply python/mxnet/ndarray/ndarray.py]: an error should be raised

File: python/mxnet/ndarray/ndarray.py [hunk 3]
```diff
@@ -968,6 +973,14 @@ def reshape(self, shape):
...
+            for key, value in kwargs.items():
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: no need to iterate. test for len(kwargs) == 1 and kwargs.get('shape', None) directly

Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: it also helps throw exception with the invalid argument name, which is consistent with numpy's behavior.

File: python/mxnet/ndarray/ndarray.py (hunk 2 above)
Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: OK

Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: Numpy throws an error too, though the error is confusing.
```
In [1]: import numpy as np

In [2]: a = np.ones((3,5))

In [3]: a.reshape(3,5,shape=(7,))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-cac6c36bb3d2> in <module>()
----> 1 a.reshape(3,5,shape=(7,))

TypeError: 'shape' is an invalid keyword argument for this function
```

File: python/mxnet/ndarray/ndarray.py [hunk 4]
```diff
@@ -968,6 +973,17 @@ def reshape(self, shape):
         array([[-1., -1., -1.],
                [-1., -1., -1.]], dtype=float32)
         """
+        if len(shape) == 1 and isinstance(shape[0], (list, tuple)):
+            shape = shape[0]
+        elif not shape:
+            for key, value in kwargs.items():
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: This is slow. Use test for len(kwargs) == 1, then kwargs.get('shape', None)

**Pull request merged**